src/app-layer-ssl.c

```diff
@@ -265,55 +270,390 @@ static void SSLSetTxDetectFlags(void *vtx, uint8_t dir, uint64_t flags)
-                                   uint32_t input_len)
+/**
+ * \inline
+ * \brief Check if value is GREASE.
+ *
+ * http://tools.ietf.org/html/draft-davidben-tls-grease-00
+ *
+ * \param value Value to check.
+ *
+ * \retval 1 if is GREASE.
+ * \retval 0 if not is GREASE.
+ */
+static inline int TLSDecodeValueIsGREASE(const uint16_t value)
 {
-    uint8_t *initial_input = input;
+    switch (value)
+    {
+        case 0x0a0a:
+        case 0x1a1a:
+        case 0x2a2a:
+        case 0x3a3a:
+        case 0x4a4a:
+        case 0x5a5a:
+        case 0x6a6a:
+        case 0x7a7a:
+        case 0x8a8a:
+        case 0x9a9a:
+        case 0xaaaa:
+        case 0xbaba:
+        case 0xcaca:
+        case 0xdada:
+        case 0xeaea:
+        case 0xfafa:
+            return 1;
+        default:
+            return 0;
+    }
+}
 
-    /* only parse the message if it is complete */
-    if (input_len < ssl_state->curr_connp->message_length || input_len < 40)
-        return 0;
+static inline int TLSDecodeHSHelloVersion(SSLState *ssl_state, uint8_t **input,
+                                          const uint32_t input_len,
+                                          const uint8_t *initial_input)
+{
+    if ((ssl_state->current_flags & SSL_AL_FLAG_STATE_CLIENT_HELLO) &&
+            ssl_config.enable_ja3) {
+        if (!(HAS_SPACE(SSLV3_CLIENT_HELLO_VERSION_LEN))) {
+            SCLogDebug("TLS handshake invalid length");
+            SSLSetEvent(ssl_state,
+                        TLS_DECODER_EVENT_HANDSHAKE_INVALID_LENGTH);
+            return -1;
+        }
 
-    /* skip version */
-    input += SSLV3_CLIENT_HELLO_VERSION_LEN;
+        uint16_t version = **input << 8 | *(*input + 1);
 
-    /* skip random */
-    input += SSLV3_CLIENT_HELLO_RANDOM_LEN;
+        ssl_state->ja3_str = Ja3BufferInit();
+        if (ssl_state->ja3_str == NULL)
+            return -1;
 
-    if (!(HAS_SPACE(1)))
-        goto invalid_length;
+        int rc = Ja3BufferAddValue(ssl_state->ja3_str, version);
+        if (rc != 0)
+            return -1;
+    }
+
+    *input += SSLV3_CLIENT_HELLO_VERSION_LEN;
+
+    return 0;
+}
+
+static inline int TLSDecodeHSHelloRandom(SSLState *ssl_state, uint8_t **input,
+                                         const uint32_t input_len,
+                                         const uint8_t *initial_input)
+{
+    /* Skip random */
+    *input += SSLV3_CLIENT_HELLO_RANDOM_LEN;
+
+    return 0;
+}
+
+static inline int TLSDecodeHSHelloSessionID(SSLState *ssl_state,
+                                            uint8_t **input,
+                                            const uint32_t input_len,
+                                            const uint8_t *initial_input)
+{
+    if (!(HAS_SPACE(1))) {
+        SCLogDebug("TLS handshake invalid length");
+        SSLSetEvent(ssl_state,
+                    TLS_DECODER_EVENT_HANDSHAKE_INVALID_LENGTH);
+        return -1;
+    }
+
+    uint8_t session_id_length = **input;
```

Comment:
I'm not too excited about this ptr-to-ptr use. This makes already fragile code even harder to track. It might be better to create a new data type that we pass around containing the ptr and size. Plus some helper funcs to safely read mem.

Reply:
I have a plan for fixing this. Don't know if you will approve, but at least you get to close another PR if you don't ;)